File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/JsonViewer.kt
package com.kinetixui.ui

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * KinetixJsonViewer — mirrors `packages/ui/src/components/json-viewer.tsx`:
 * a collapsible, syntax-colored tree over arbitrary JSON-shaped data.
 * Accepts a plain `Map<String, Any?>` / `List<Any?>` / primitive tree — the
 * ad-hoc shape most JSON deserializes into without pulling in a specific
 * serialization library — rather than a sealed `JsonElement` type. Unlike
 * [KinetixDataGrid], a recursive expand/collapse tree needs no gesture or
 * layout primitive native platforms lack, so this ships with the same
 * feature set as web.
 */
@Composable
fun KinetixJsonViewer(
    value: Any?,
    modifier: Modifier = Modifier,
    expandDepth: Int = 1,
) {
    Column(modifier = modifier) {
        JsonNode(name = null, value = value, depth = 0, expandDepth = expandDepth, isLast = true)
    }
}

@Composable
private fun JsonNode(name: String?, value: Any?, depth: Int, expandDepth: Int, isLast: Boolean) {
    val colors = KinetixColorScheme.current
    val isContainer = value is Map<*, *> || value is List<*>

    if (!isContainer) {
        Row(modifier = Modifier.padding(start = (depth * 16).dp)) {
            if (name != null) {
                Text("\"$name\": ", color = colors.mutedForeground, fontSize = 12.sp)
            }
            JsonPrimitiveText(value)
            if (!isLast) Text(",", color = colors.mutedForeground, fontSize = 12.sp)
        }
        return
    }

    val isArrayVal = value is List<*>
    val entries: List<Pair<String, Any?>> = when (value) {
        is List<*> -> value.mapIndexed { i, v -> i.toString() to v }
        is Map<*, *> -> value.entries.map { it.key.toString() to it.value }
        else -> emptyList()
    }
    val openBracket = if (isArrayVal) "[" else "{"
    val closeBracket = if (isArrayVal) "]" else "}"

    if (entries.isEmpty()) {
        Row(modifier = Modifier.padding(start = (depth * 16 + 20).dp)) {
            if (name != null) Text("\"$name\": ", color = colors.mutedForeground, fontSize = 12.sp)
            Text("$openBracket$closeBracket", fontSize = 12.sp)
            if (!isLast) Text(",", color = colors.mutedForeground, fontSize = 12.sp)
        }
        return
    }

    var expanded by remember { mutableStateOf(depth < expandDepth) }

    Column {
        Row(
            modifier = Modifier
                .padding(start = (depth * 16).dp)
                .clickable { expanded = !expanded },
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                "▸",
                fontSize = 10.sp,
                color = colors.mutedForeground,
                modifier = Modifier
                    .padding(end = 4.dp)
                    .rotate(if (expanded) 90f else 0f),
            )
            if (name != null) Text("\"$name\": ", color = colors.mutedForeground, fontSize = 12.sp)
            Text(openBracket, fontSize = 12.sp)
            if (!expanded) {
                Text(
                    " ${entries.size} ${if (isArrayVal) "item" else "key"}${if (entries.size == 1) "" else "s"} ",
                    color = colors.mutedForeground,
                    fontSize = 12.sp,
                )
                Text(closeBracket, fontSize = 12.sp)
                if (!isLast) Text(",", color = colors.mutedForeground, fontSize = 12.sp)
            }
        }
        if (expanded) {
            entries.forEachIndexed { i, (key, v) ->
                JsonNode(
                    name = if (isArrayVal) null else key,
                    value = v,
                    depth = depth + 1,
                    expandDepth = expandDepth,
                    isLast = i == entries.lastIndex,
                )
            }
            Row(modifier = Modifier.padding(start = (depth * 16).dp)) {
                Text(closeBracket, fontSize = 12.sp)
                if (!isLast) Text(",", color = colors.mutedForeground, fontSize = 12.sp)
            }
        }
    }
}

@Composable
private fun JsonPrimitiveText(value: Any?) {
    val colors = KinetixColorScheme.current
    when (value) {
        null -> Text("null", color = colors.mutedForeground, fontSize = 12.sp, fontStyle = FontStyle.Italic)
        is String -> Text("\"$value\"", color = colors.success, fontSize = 12.sp)
        is Number -> Text(value.toString(), color = colors.info, fontSize = 12.sp)
        is Boolean -> Text(value.toString(), color = colors.warning, fontSize = 12.sp)
        else -> Text(value.toString(), fontSize = 12.sp)
    }
}

File: packages/ui-flutter/lib/src/json_viewer.dart
import 'package:flutter/material.dart';

import 'theme.dart';

/// Mirrors `packages/ui/src/components/json-viewer.tsx`: a collapsible,
/// syntax-colored tree over arbitrary JSON-shaped data. Accepts the same
/// shape `dart:convert`'s `jsonDecode` produces — `Map<String, dynamic>`,
/// `List<dynamic>`, `String`, `num`, `bool`, or `null` — directly, no
/// wrapper type needed the way `KinetixJSONValue` is on SwiftUI. Unlike
/// `KinetixDataGrid`, a recursive expand/collapse tree needs no gesture or
/// layout primitive Flutter lacks, so this ships with the same feature set
/// as web.
class KinetixJsonViewer extends StatelessWidget {
  const KinetixJsonViewer({super.key, required this.data, this.expandDepth = 1});

  final dynamic data;
  final int expandDepth;

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      child: Padding(
        padding: const EdgeInsets.all(8),
        child: _JsonNode(name: null, value: data, depth: 0, expandDepth: expandDepth, isLast: true),
      ),
    );
  }
}

class _JsonNode extends StatefulWidget {
  const _JsonNode({
    required this.name,
    required this.value,
    required this.depth,
    required this.expandDepth,
    required this.isLast,
  });

  final String? name;
  final dynamic value;
  final int depth;
  final int expandDepth;
  final bool isLast;

  @override
  State<_JsonNode> createState() => _JsonNodeState();
}

class _JsonNodeState extends State<_JsonNode> {
  late bool _expanded = widget.depth < widget.expandDepth;

  static const _fontSize = 12.0;

  TextStyle _mono(Color color, {bool italic = false}) => TextStyle(
        fontFamily: 'monospace',
        fontSize: _fontSize,
        color: color,
        fontStyle: italic ? FontStyle.italic : FontStyle.normal,
      );

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);
    final value = widget.value;
    final isContainer = value is Map || value is List;

    if (!isContainer) {
      return Padding(
        padding: EdgeInsets.only(left: (widget.depth * 16).toDouble()),
        child: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            if (widget.name != null) Text('"${widget.name}": ', style: _mono(c.mutedForeground)),
            _primitive(c, value),
            if (!widget.isLast) Text(',', style: _mono(c.mutedForeground)),
          ],
        ),
      );
    }

    final isArray = value is List;
    final entries = isArray
        ? List.generate((value as List).length, (i) => MapEntry(i.toString(), value[i]))
        : (value as Map).entries.map((e) => MapEntry(e.key.toString(), e.value)).toList();
    final openBracket = isArray ? '[' : '{';
    final closeBracket = isArray ? ']' : '}';

    if (entries.isEmpty) {
      return Padding(
        padding: EdgeInsets.only(left: (widget.depth * 16 + 20).toDouble()),
        child: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            if (widget.name != null) Text('"${widget.name}": ', style: _mono(c.mutedForeground)),
            Text('$openBracket$closeBracket', style: _mono(c.foreground)),
            if (!widget.isLast) Text(',', style: _mono(c.mutedForeground)),
          ],
        ),
      );
    }

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: [
        Padding(
          padding: EdgeInsets.only(left: (widget.depth * 16).toDouble()),
          child: GestureDetector(
            behavior: HitTestBehavior.opaque,
            onTap: () => setState(() => _expanded = !_expanded),
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                AnimatedRotation(
                  turns: _expanded ? 0.25 : 0,
                  duration: const Duration(milliseconds: 100),
                  child: Icon(Icons.chevron_right, size: 14, color: c.mutedForeground),
                ),
                if (widget.name != null) Text('"${widget.name}": ', style: _mono(c.mutedForeground)),
                Text(openBracket, style: _mono(c.foreground)),
                if (!_expanded) ...[
                  Text(
                    ' ${entries.length} ${isArray ? 'item' : 'key'}${entries.length == 1 ? '' : 's'} ',
                    style: _mono(c.mutedForeground),
                  ),
                  Text(closeBracket, style: _mono(c.foreground)),
                  if (!widget.isLast) Text(',', style: _mono(c.mutedForeground)),
                ],
              ],
            ),
          ),
        ),
        if (_expanded) ...[
          for (var i = 0; i < entries.length; i++)
            _JsonNode(
              name: isArray ? null : entries[i].key,
              value: entries[i].value,
              depth: widget.depth + 1,
              expandDepth: widget.expandDepth,
              isLast: i == entries.length - 1,
            ),
          Padding(
            padding: EdgeInsets.only(left: (widget.depth * 16).toDouble()),
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                Text(closeBracket, style: _mono(c.foreground)),
                if (!widget.isLast) Text(',', style: _mono(c.mutedForeground)),
              ],
            ),
          ),
        ],
      ],
    );
  }

  Widget _primitive(KinetixColors c, dynamic value) {
    if (value == null) return Text('null', style: _mono(c.mutedForeground, italic: true));
    if (value is String) return Text('"$value"', style: _mono(c.success));
    if (value is num) return Text(value.toString(), style: _mono(c.info));
    if (value is bool) return Text(value.toString(), style: _mono(c.warning));
    return Text(value.toString(), style: _mono(c.foreground));
  }
}
